### PR TITLE
Modify script size validation when tx is segwit. Add tests from real transactions

### PR DESCRIPTION
**Pull request by julia-zack**

### CODE REVIEW

File: rskj-core/src/main/java/co/rsk/peg/bitcoin/BitcoinUtils.java [hunk 1]
```diff
@@ -179,11 +178,28 @@ public static void addSpendingFederationBaseScript(BtcTransaction btcTx, int inp
     }
 
     private static void setSegwitScriptSig(TransactionInput txIn, Script redeemScript) {
+        Script segwitScriptSig = buildSegwitScriptSig(redeemScript);
+        txIn.setScriptSig(segwitScriptSig);
+    }
```
Comment (nathanieliov) [inline rskj-core/src/main/java/co/rsk/peg/bitcoin/BitcoinUtils.java:183]: I think you don't need this method.  You should call `buildSegwitScriptSig` from `addSpendingFederationBaseScript`. Imo, this method does not add any value.

File: rskj-core/src/test/java/co/rsk/peg/bitcoin/BitcoinUtilsTest.java [hunk 2]
```diff
@@ -1152,6 +1143,57 @@ void addSpendingFederationBaseScript_forSegwitFed_shouldAddExpectedInputWitness(
...
+        assertEquals("20", opPushBytes32);
+
+        // hashed redeem script
+        byte[] hashedRedeemScript = extractHashedRedeemScriptProgramFromSegwitScriptSig(segwitScriptSig);
+        byte[] expectedHashedRedeemScript = Sha256Hash.hash(federation.getRedeemScript().getProgram());
+        assertArrayEquals(expectedHashedRedeemScript, hashedRedeemScript);
+    }
+
+    @Test
+    void extractHashedRedeemScriptProgramFromSegwitScriptSig_returnsExpectedHash() {
```
Comment (nathanieliov) [inline rskj-core/src/test/java/co/rsk/peg/bitcoin/BitcoinUtilsTest.java:1184]: Missing unit test for NonSegwitScriptSig, Empty, null.

File: rskj-core/src/test/java/co/rsk/peg/bitcoin/BitcoinUtilsTest.java [hunk 3]
```diff
@@ -1152,6 +1143,57 @@ void addSpendingFederationBaseScript_forSegwitFed_shouldAddExpectedInputWitness(
         assertEquals(oneChunkSegwitScriptSig, transaction.getInput(inputIndex).getScriptSig());
     }
 
+    @Test
+    void buildSegwitScriptSig_buildsScriptSigWithExpectedStructure() {
```
Comment (nathanieliov) [inline rskj-core/src/test/java/co/rsk/peg/bitcoin/BitcoinUtilsTest.java:1147]: Missing unit tests for NonSegwitScriptSig, Empty, null.

File: rskj-core/src/main/java/co/rsk/peg/bitcoin/ScriptValidations.java [hunk 4]
```diff
@@ -7,7 +7,8 @@
 public class ScriptValidations {
     private ScriptValidations() {
     }
-    public static void validateScriptSize(Script script) throws ScriptCreationException {
+
+    public static void validateScriptSigElementSize(Script script) throws ScriptCreationException {
```
[the remark this answers is not in the document]
Comment (marcos-iov) [reply rskj-core/src/main/java/co/rsk/peg/bitcoin/ScriptValidations.java:11]: I don't think the new name is a very good idea. The restriction of 520 bytes max applies to redeem scripts if I'm not mistaken. And that is what we are validating here, not scriptSigs

File: rskj-core/src/main/java/co/rsk/peg/bitcoin/P2shErpRedeemScriptBuilder.java [hunk 5]
```diff
@@ -33,12 +33,8 @@ public Script of(
 
         byte[] serializedCsvValue = Utils.signedLongToByteArrayLE(csvValue);
         logger.debug("[createRedeemScriptFromKeys] Creating the redeem script from the scripts");
-        Script redeemScript = createRedeemScriptFromScripts(defaultRedeemScript, emergencyRedeemScript, serializedCsvValue);
 
-        logger.debug("[createRedeemScriptFromKeys] Validating redeem script size");
-        ScriptValidations.validateScriptSize(redeemScript);
```
[the remark this answers is not in the document]
Comment (marcos-iov) [reply rskj-core/src/main/java/co/rsk/peg/bitcoin/P2shErpRedeemScriptBuilder.java:39]: P2wshP2sh allows for multisigs of up to 20 signers. It uses this class to build the standard multisig part

File: rskj-core/src/main/java/co/rsk/peg/federation/ErpFederation.java [hunk 6]
```diff
@@ -74,13 +75,26 @@ public Script getRedeemScript() {
...
 
+    private void validateScriptSize() {
+        if (getFormatVersion() != P2SH_P2WSH_ERP_FEDERATION.getFormatVersion()) {
+            // since the redeem script is located in the script sig,
+            // we need to check it does not surpass the maximum allowed size
+            ScriptValidations.validateScriptSigElementSize(redeemScript);
+            return;
+        }
+
+        ScriptValidations.validateWitnessScriptSize(redeemScript, getNumberOfSignaturesRequired());
```
Comment (marcos-iov) [inline rskj-core/src/main/java/co/rsk/peg/federation/ErpFederation.java:95]: Are we sure the limit is impossed to the witness script or is it to the redeem script?

File: rskj-core/src/main/java/co/rsk/peg/bitcoin/BitcoinUtils.java (hunk 1 above)
Comment (jeremy-then) [reply rskj-core/src/main/java/co/rsk/peg/bitcoin/BitcoinUtils.java]: Updated.

File: rskj-core/src/test/java/co/rsk/peg/bitcoin/BitcoinUtilsTest.java (hunk 2 above)
Comment (jeremy-then) [reply rskj-core/src/test/java/co/rsk/peg/bitcoin/BitcoinUtilsTest.java:1233]: Updated.

File: rskj-core/src/test/java/co/rsk/peg/bitcoin/BitcoinUtilsTest.java (hunk 3 above)
Comment (jeremy-then) [reply rskj-core/src/test/java/co/rsk/peg/bitcoin/BitcoinUtilsTest.java:1147]: Updated.

File: rskj-core/src/main/java/co/rsk/peg/bitcoin/ScriptValidations.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (jeremy-then) [reply rskj-core/src/main/java/co/rsk/peg/bitcoin/ScriptValidations.java]: Updated.

File: rskj-core/src/main/java/co/rsk/peg/bitcoin/ScriptValidations.java [hunk 7]
```diff
@@ -17,4 +18,18 @@ public static void validateScriptSize(Script script) throws ScriptCreationExcept
             throw new ScriptCreationException(message, ABOVE_MAX_SCRIPT_ELEMENT_SIZE);
         }
     }
+
+    public static void validateWitnessScriptSize(Script redeemScript, int numberOfSignaturesRequired) throws ScriptCreationException {
```
[the remark this answers is not in the document]
Comment (jeremy-then) [reply rskj-core/src/main/java/co/rsk/peg/bitcoin/ScriptValidations.java]: Updated.

File: rskj-core/src/main/java/co/rsk/peg/bitcoin/BitcoinUtils.java [hunk 8]
```diff
@@ -179,11 +178,28 @@ public static void addSpendingFederationBaseScript(BtcTransaction btcTx, int inp
...
+            .data(segwitScriptSig.getProgram())
+            .build();
+    }
 
-        txIn.setScriptSig(oneChunkSegwitScriptSig);
+    public static byte[] extractHashedRedeemScriptProgramFromSegwitScriptSig(Script segwitScriptSig) {
+        byte[] segwitScriptSigProgram = segwitScriptSig.getProgram();
+        // the whole program is [22 00 20 + rs]
+        String hashedRedeemScript = Hex.toHexString(segwitScriptSigProgram).substring(6);
+        return Hex.decode(hashedRedeemScript);
```
[the remark this answers is not in the document]
Comment (jeremy-then) [reply rskj-core/src/main/java/co/rsk/peg/bitcoin/BitcoinUtils.java]: Updated.

File: rskj-core/src/main/java/co/rsk/peg/federation/ErpFederation.java (hunk 6 above)
Comment (jeremy-then) [reply rskj-core/src/main/java/co/rsk/peg/federation/ErpFederation.java]: Updated.

File: rskj-core/src/main/java/co/rsk/peg/bitcoin/BitcoinUtils.java [hunk 9]
```diff
@@ -196,10 +195,11 @@ public static Script buildSegwitScriptSig(Script redeemScript) {
     }
 
     public static byte[] extractHashedRedeemScriptProgramFromSegwitScriptSig(Script segwitScriptSig) {
+        if (segwitScriptSig == null) {
```
[the remark this answers is not in the document]
Comment (jeremy-then) [reply rskj-core/src/main/java/co/rsk/peg/bitcoin/BitcoinUtils.java:198]: Not really. If the size is 0, you will get an empty array as well. No out of bounds exception thrown.
